### PR TITLE
[4.1][stdlib] Eliminate IndexDistance

### PR DESCRIPTION
**Pull request by airspeedswift**
- **Explanation**: Eliminate the `IndexDistance` associated type, implementing [SE-191](https://github.com/apple/swift-evolution/blob/master/proposals/0191-eliminate-indexdistance.md).
- **Scope**: This touches a large number of types in the standard library, simplifying them to replace all occurrences of `IndexDistance` with `Int`. It also introduces some compatibility shims (an `IndexDistance` typealias equal to `Int`, and generic conversions for calls using non-`Int` index distances). 
- **Issue**: rdar://problem/35940300
- **Reviewed by**: @moiseev 
- **Risk**: Low; this is a pervasive change on all collections but the changes are fairly mechanical and easily reviewed. However, there are no known uses where `Int` is not already used, and the majority of changes are simplifications of the code to no longer need to perform casts between different distance types.
- **Testing**: The existing collection implementations in the standard library are extensively tested. Additionally this passes the compatibility suite, and also passes tests of other collections implemented in Foundation and SwiftPM.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by airspeedswift** [comment]
@swift-ci please smoke compiler performance

File: stdlib/public/core/Collection.swift [hunk 1]
```diff
@@ -782,6 +776,9 @@ public protocol Collection: Sequence where SubSequence: Collection {
   /// - Parameter i: A valid index of the collection. `i` must be less than
   ///   `endIndex`.
   func formIndex(after i: inout Index)
+
+  @available(swift, deprecated, message: "all index distances are now of type Int")
+  typealias IndexDistance = Int
```
Comment (moiseev) [inline stdlib/public/core/Collection.swift]: `public`.

File: stdlib/public/core/Collection.swift [hunk 2]
```diff
@@ -1701,9 +1698,30 @@ extension Collection {
...
+  public func formIndex<T: BinaryInteger>(_ i: Index, offsetBy n: T) {
+    return formIndex(i, offsetBy: Int(n))
+  }
+  @available(*, deprecated, message: "all index distances are now of type Int")
+  public func index<T: BinaryInteger>(_ i: Index, offsetBy n: T, limitedBy limit: Index) -> Index {
+    return index(i, offsetBy: Int(n), limitedBy: limit)
+  }
+  */
+  @available(*, deprecated, message: "all index distances are now of type Int")
+  public func formIndex<T: BinaryInteger>(_ i: inout Index, offsetBy n: T, limitedBy limit: Index) -> Bool {
```
Comment (moiseev) [inline stdlib/public/core/Collection.swift]: I think `formIndex` should have a `@discardableResult` attribute, otherwise it will issue 2 warnings.

File: stdlib/public/core/Collection.swift [hunk 3]
```diff
@@ -1701,9 +1698,30 @@ extension Collection {
...
+    return formIndex(&i, offsetBy: Int(n), limitedBy: limit)
+  }
+  @available(*, deprecated, message: "all index distances are now of type Int")
+  public func distance<T: BinaryInteger>(from start: Index, to end: Index) -> T {
```
Comment (moiseev) [inline stdlib/public/core/Collection.swift]: Surprised it does not result in ambiguity in expressions like `let d = xs.distance(from: someIndex, to: someOtherIndex)`.

File: stdlib/public/core/Integers.swift.gyb [hunk 4]
```diff
@@ -2898,6 +2898,9 @@ public struct ${Self}
   : FixedWidthInteger, ${Unsigned}Integer,
     _ExpressibleByBuiltinIntegerLiteral {
 
+  public typealias IntegerLiteralType = ${Self}
```
Comment (moiseev) [inline stdlib/public/core/Integers.swift.gyb]: How is this related?

File: stdlib/public/core/Collection.swift (hunk 1 above)
Comment (airspeedswift) [reply stdlib/public/core/Collection.swift]: Yup. I moved it back and forth between the protocol and the extension multiple times and forgot this every. single. time.

File: stdlib/public/core/Collection.swift (hunk 3 above)
Comment (airspeedswift) [reply stdlib/public/core/Collection.swift]: The other `distance` method isn't generic (`T` is `Int`) so it wins.

File: stdlib/public/core/Integers.swift.gyb (hunk 4 above)
Comment (airspeedswift) [reply stdlib/public/core/Integers.swift.gyb]: Hum. It isn't. I probably meant to do this on my clean-up branch. But since this is a cherry-pick from master where it's merged, probably should just leave it in.

File: stdlib/public/core/Collection.swift (hunk 3 above)
Comment (airspeedswift) [reply stdlib/public/core/Collection.swift]: Isn't generic other than on the protocol, that is...

**Comment by airspeedswift** [comment]
@swift-ci please nominate

File: stdlib/public/core/Collection.swift (hunk 2 above)
Comment (airspeedswift) [reply stdlib/public/core/Collection.swift]: eh, they're both legit warnings though

**Review by moiseev — APPROVED** [review]

File: stdlib/public/core/Collection.swift (hunk 2 above)
Comment (moiseev) [reply stdlib/public/core/Collection.swift]: Take that back. Turns out we do not have `@discardableResult` on this method anywhere...
